### PR TITLE
Auth totp

### PR DESCRIPTION
**Pull request by powerful23**
*Issue #, if available:*
Supporting TOTP for #215 #164 #111 
*Description of changes:*
Adding TOTP related methods in Auth class as well as UI components into ```Authenticator``` to setup TOTP. Also Adding UI components for choosing MFA types. Note: the UI components are only for React

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by powerful23** [comment]
You can skip reviewing the last two commits which is only for unit tests.

File: packages/aws-amplify-react/__tests__/Storage/__snapshots__/S3Album-test.js.snap [hunk 1]
```diff
@@ -494,7 +494,11 @@ exports[`S3Album test render test render correctly if has images and texts 1`] =
   </div>
   <Picker
     accept="image/*, text/*"
+<<<<<<< HEAD
+    key="1520285142742"
+=======
     key="1520027643564"
+>>>>>>> upstream/master
```
Comment (mlabieniec) [inline packages/aws-amplify-react/__tests__/Storage/__snapshots__/S3Album-test.js.snap]: Looks like a conflict here

File: packages/aws-amplify/src/Auth/Auth.ts [hunk 2]
```diff
@@ -244,6 +246,201 @@ export default class AuthClass {
...
+                mfaSetup: (challengeName, challengeParam) => {
+                    logger.debug('signIn mfa setup', challengeName);
+                    user['challengeName'] = challengeName;
+                    user['challengeParam'] = challengeParam;
+                    resolve(user);
+                },
+                // associateSecretCode: (secretCode) => {
+                //     logger.debug('signIn asscoiateSecretCode', secretCode);
+                //     resolve(user);
+                // },
```
Comment (mlabieniec) [inline packages/aws-amplify/src/Auth/Auth.ts]: should we remove this commented code?

File: packages/aws-amplify/src/Auth/Auth.ts [hunk 3]
```diff
@@ -244,6 +246,201 @@ export default class AuthClass {
...
+
+        const that = this;
+        const TOTP_NOT_VERIFED = 'User has not verified software token mfa';
+        const TOTP_NOT_SETUPED = 'User has not set up software token mfa';
+        return new Promise((res, rej) => {
+            user.setUserMfaPreference(smsMfaSettings, totpMfaSettings, (err, result) => {
+                if (err) {
+                    // if totp not setup or verified and user want to set it, return error
+                    // otherwise igonre it
+                    if (err.message === TOTP_NOT_SETUPED || err.message === TOTP_NOT_VERIFED) {
```
Comment (mlabieniec) [inline packages/aws-amplify/src/Auth/Auth.ts]: is this correct? "TOTP_NOT_SETUPED " Should it be "TOTP_NOT_SETUP" ?

**Pull request merged**